<div style="color:white; background-color:#263b52; text-align:center; padding: 25px 0;">
    <div style="font-size:34px; font-family:calibri; font-weight:bold;">
        ML Zoomcamp 2025
    </div>
    <div style="font-size:22px; font-family:verdana; line-height: 1.5; margin-top:10px;">
        Homework 1: Introduction to ML - Environment & Pandas
    </div>
</div>

<div style="margin-top:20px; overflow:auto;">
  <div style="float:left; width:50%; text-align:left; font-family:verdana; font-size:16px; color:#263b52;">
      <p style="margin:4px 0;"><b>Author:</b> CGD</p>
      <p style="margin:4px 0;"><b>GitHub:</b> <a href="https://github.com/CGD2401" target="_blank">CGD2401</a></p>
      <p style="margin:4px 0;"><b>Date:</b> Septiembre 2025</p>
  </div>

  <div style="float:right; width:50%; text-align:right; font-family:verdana; font-size:16px; color:#263b52;">
      <p style="margin:4px 0;">Notebook prepared for</p>
      <p style="margin:4px 0;"><b>Machine Learning Zoomcamp</b></p>
      <p style="margin:4px 0;">Homework covering Q1–Q7</p>
  </div>
</div>

<div style="width:100%; clear:both;"></div>

<hr style="border:3px solid #7ba5b0; margin-top:25px;"/>


## 1. Importing Libraries
In this section, we load the essential Python libraries required for the homework.  
- **NumPy** for numerical operations  
- **Pandas** for data manipulation and analysis  
- **Matplotlib** and **Seaborn** for data visualization 

In [4]:
# Core libraries
import numpy as np              # Numerical operations
import pandas as pd             # Data analysis and manipulation

# Visualization
import matplotlib.pyplot as plt # Basic plotting
import seaborn as sns           # Advanced statistical plots

# Display versions for reproducibility
print("NumPy version:", np.__version__)
print("Pandas version:", pd.__version__)

NumPy version: 1.26.4
Pandas version: 2.2.3


## 2. Loading the Dataset

In this section we download and load the Car Fuel Efficiency dataset into a Pandas DataFrame.

In [5]:
url = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv"
df = pd.read_csv(url)
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

## 3. Exploratory Data Analysis (EDA)

In this section we perform basic exploration to understand the dataset’s structure and quality.

In [6]:
df.info()
df.describe()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

## 4. Homework Questions

### Q1. Pandas Version

In [7]:
pd.__version__

'2.2.3'

### Q2. Records Count

In [8]:
len(df)

9704

### Q3. Fuel Types

In [9]:
df['fuel_type'].nunique()

2

### Q4. Missing Values

In [10]:
df.isnull().any().sum()

4

### Q5. Maximum Fuel Efficiency (Asia)

In [11]:
print(df.columns.tolist())

['engine_displacement', 'num_cylinders', 'horsepower', 'vehicle_weight', 'acceleration', 'model_year', 'origin', 'fuel_type', 'drivetrain', 'num_doors', 'fuel_efficiency_mpg']


In [12]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max().round(3)


23.759

### Q6. Median value of Horsepower (before and after filling missing values)

In [13]:
median_before = df['horsepower'].median()
print("Median before filling:", median_before)

Median before filling: 149.0


In [14]:
hp_mode = df['horsepower'].mode()[0]
print("Most frequent (mode):", hp_mode)

Most frequent (mode): 152.0


In [15]:
df['horsepower'] = df['horsepower'].fillna(hp_mode)

In [16]:
median_after = df['horsepower'].median()
print("Median after filling:", median_after)

Median after filling: 152.0


In [17]:
if median_after > median_before:
    print("Yes, it increased")
elif median_after < median_before:
    print("Yes, it decreased")
else:
    print("No")

Yes, it increased


### Q7. Sum of Weights - Linear Regression Warm-up

In [18]:
X = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7).to_numpy()
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

XTX = X.T @ X
XTX_inv = np.linalg.inv(XTX)
w = XTX_inv @ X.T @ y
w.sum().round(3)

0.519

### 5. Conclusions


In this homework, we set up the Python environment with NumPy, Pandas, Matplotlib, and Seaborn, confirming reproducibility with NumPy 1.26.4 and Pandas 2.2.3.

The Car Fuel Efficiency dataset was successfully loaded from the GitHub source, containing variables such as engine displacement, horsepower, vehicle weight, model year, origin, and fuel efficiency in miles per gallon. Through exploratory analysis, we observed that the dataset is rich and diverse but also contains missing values, particularly in columns such as `horsepower` and `num_cylinders`, which required careful handling.

Answering the homework questions deepened our understanding of basic data analysis techniques. We confirmed the dataset size, the distribution of fuel types, and the presence of missing values. For cars originating from Asia, we identified the maximum fuel efficiency. When addressing missing values in the horsepower column, we replaced them with the most frequent value and observed that the **median increased** as a result.

Finally, we implemented the mathematical steps of linear regression manually with NumPy, obtaining a sum of regression weights close to 0.519.

Overall, this exercise highlighted the importance of reproducibility, exploratory data analysis, and proper handling of missing values. It also provided a hands-on introduction to the linear algebra foundations of machine learning, setting the stage for more advanced concepts in the upcoming lessons of the ML Zoomcamp.